torch/csrc/distributed/autograd/context/dist_autograd_context.cpp

```diff
@@ -48,6 +48,68 @@ DistAutogradContext::recvFunctions() const {
   return recvAutogradFunctions_;
 }
 
+void DistAutogradContext::accumulateGrad(
+    torch::autograd::Variable variable,
+    torch::Tensor grad) {
+  TORCH_INTERNAL_ASSERT(grad.defined());
+  TORCH_INTERNAL_ASSERT(variable.requires_grad());
+
+  std::lock_guard<std::mutex> guard(lock_);
+  auto it = accumulatedGrads_.find(variable);
+  if (it != accumulatedGrads_.end()) {
+    // Accumulate multiple grads on the same variable.
+    it->setValue(it->value() + grad);
+  } else {
+    // First grad for this variable.
+    accumulatedGrads_.insert(variable, grad);
+  }
+}
+
+torch::autograd::GraphTask& DistAutogradContext::retrieveGraphTask() {
+  std::lock_guard<std::mutex> guard(lock_);
+  TORCH_INTERNAL_ASSERT(graphTask_);
+  return *graphTask_;
+}
+
+void DistAutogradContext::setGraphTask(
+    std::unique_ptr<torch::autograd::GraphTask> graphTask) {
+  std::lock_guard<std::mutex> guard(lock_);
+  graphTask_ = std::move(graphTask);
+}
+
+void DistAutogradContext::addOutStandingRpc(
+    std::shared_ptr<rpc::FutureMessage> futureMessage) {
+  std::lock_guard<std::mutex> guard(lock_);
+  outStandingRpcs_.push_back(futureMessage);
+}
+
+void DistAutogradContext::waitForOutStandingRpcs() {
+  // Copy futures under lock, but wait for them outside the lock.
+  std::unique_lock<std::mutex> lock(lock_);
+  auto outStandingRpcs = outStandingRpcs_;
+  lock.unlock();
+
+  for (const auto& outStandingRpc : outStandingRpcs) {
+    outStandingRpc->wait();
+  }
+}
+
+std::shared_ptr<SendRpcBackward> DistAutogradContext::retrieveSendFunction(
+    int64_t autograd_message_id) {
+  std::lock_guard<std::mutex> guard(lock_);
+  auto it = sendAutogradFunctions_.find(autograd_message_id);
+  TORCH_CHECK(
+      it != sendAutogradFunctions_.end(),
+      "Could not find send function for autograd message id: ",
+      autograd_message_id);
+  return it->second;
+}
+
+const c10::Dict<torch::Tensor, torch::Tensor>& DistAutogradContext::
```

Comment:
I have to admit I'm a little confused about the locking discipline on this class. When do I have to take out a lock, and when is it OK to not take a lock (as is done here?)

Reply:
Good point, for now I'll just use a copy here. We probably need a more efficient way of doing this and we can later on provide a lockless `unsafeGetGradients` if the user knows that we're done with the backwards pass.